src/main/scala/com/fulcrumgenomics/bam/SplitBam.scala

```diff
@@ -55,77 +55,101 @@ object SplitType extends FgBioEnum[SplitType] {
     |The resulting BAMs will be named `<output-prefix>.<read-group-id>.bam`, or `<output-prefix>.<library-name>.bam`
     |when splitting by the library.  All reads without a read group, or without a library when splitting by library,
     |will be written to `<output-prefix>.unknown.bam`.  If no such reads exist, then no such file will exist.
+    |
+    |For splitting a BAM with many read groups, use `--no-async-writing` in case too many threads or memory are used.
   """)
 class SplitBam
 ( @arg(flag='i', doc="Input SAM or BAM file.") val input: PathToBam,
   @arg(flag='o', doc="Output prefix for all SAM or BAM files (ex. output/sample-name).") val output: PathPrefix,
   @arg(flag='s', doc="Split by library instead of read group") val splitBy: SplitType = SplitType.ReadGroup,
-  @arg(flag='u', doc="The name to use for the unknown file") val unknown: String = "unknown"
+  @arg(flag='u', doc="The name to use for the unknown file") val unknown: String = "unknown",
+  @arg(doc="Do not write the records asynchronously. Use this if with a large numbers of read groups to reduce memory usage.")
+  val noAsyncWriting: Boolean = false
 ) extends FgBioTool with LazyLogging {
 
   Io.assertReadable(input)
   Io.assertCanWriteFile(output)
 
   override def execute(): Unit = {
     val in       = SamSource(input)
-    val progress = ProgressLogger(logger)
-    val unknownBamAndWriter = {
-      val unknownBam: PathToBam = toOutput(unknown)
-      val unknownWriter: SamWriter = SamWriter(toOutput(unknown), in.header)
-      //new SAMFileWriterFactory().makeWriter(in.h, true, toOutput(unknown).toFile, null)
+    val progress = ProgressLogger(logger=logger)
+    val unknownBamAndWriter: WriterInfo = {
+      val unknownBam: PathToBam = toOutput(name=unknown)
+      val unknownWriter: SamWriter = SamWriter(toOutput(unknown), toOutputHeader(in.header))
       WriterInfo(name=unknown, bam=unknownBam, writer=unknownWriter)
     }
-    val writers  = createWriters(header=in.header, splitBy=splitBy).withDefaultValue(unknownBamAndWriter)
-    val counter  = new SimpleCounter[WriterInfo]()
+    val writerInfoMap: Map[SAMReadGroupRecord, WriterInfo] = {
+      createWriters(header=in.header, splitBy=splitBy).withDefaultValue(unknownBamAndWriter)
+    }
 
     in.foreach { rec =>
-      val info = writers(rec.readGroup)
-      info.writer += rec
-      counter.count(info)
+      val info = writerInfoMap(rec.readGroup)
+      info.count += 1
+      info.writer.write(rec)
       progress.record(rec)
     }
+    progress.logLast()
 
-    writers.values.foreach(_.close())
-    unknownBamAndWriter.close()
     in.safelyClose()
 
-    counter.toSeq sortBy { _._1.name } foreach { case (info, count) =>
-      logger.info(s"Wrote $count records to ${info.bam.getFileName}")
+    writerInfoMap.values.toSeq.sortBy(_.name).distinct.foreach { info =>
```

Comment:
`.distinct` is added for good measure in case.  There could be multiple `WriterInfo`s that write to the same output (ex. `SplitByLibrary`)